crates/mongodb-connector/src/mongo_connector.rs

```diff
@@ -141,18 +141,27 @@ impl Connector for MongoConnector {
         state: &Self::State,
         request: QueryRequest,
     ) -> Result<JsonResponse<QueryResponse>, QueryError> {
-        tracing::debug!(query_request = %serde_json::to_string(&request).unwrap(), "received query request");
-        let query_context = get_query_context(configuration);
-        let v2_request = v3_to_v2_query_request(&query_context, request.clone())?;
-        let response_documents = handle_query_request(configuration, state, v2_request)
-            .await
-            .map_err(mongo_agent_error_to_query_error)?;
-        let response = serialize_query_response(&query_context, &request, response_documents)
-            .map_err(|err| {
-                QueryError::UnprocessableContent(format!(
-                    "error converting MongoDB response to JSON: {err}"
-                ))
+        let response = async move {
+            tracing::debug!(query_request = %serde_json::to_string(&request).unwrap(), "received query request");
+            let query_context = get_query_context(configuration);
+            let v2_request = tracing::info_span!("Prepare Query Request").in_scope(|| {
+                v3_to_v2_query_request(&query_context, request.clone())
             })?;
+            let response_documents = handle_query_request(configuration, state, v2_request)
+                .instrument(tracing::info_span!("Process Query Request", internal.visibility = "user"))
+                .await
+                .map_err(mongo_agent_error_to_query_error)?;
+            tracing::info_span!("Serialize Query Response", internal.visibility = "user").in_scope(|| {
+                serialize_query_response(&query_context, &request, response_documents)
+                .map_err(|err| {
+                    QueryError::UnprocessableContent(format!(
+                        "error converting MongoDB response to JSON: {err}"
+                    ))
+                })
+            })
+        }
+        .instrument(tracing::info_span!("/query", internal.visibility = "user"))
```

Comment:
Since this instrumentation applies to the effectively the entire function, I think another way to do this would be to modify the `#[instrument]` annotation on the function. Currently it's configured to only emit traces on errors, but it can be configured to do other things. One thing to watch out for is that using `#[instrument]` on a function captures argument values unless you opt out - the `skip_all` option excludes all arguments from the trace.

But this way works just as well.